epan/dissectors/packet-ssh.c

```diff
@@ -370,17 +554,40 @@ dissect_ssh(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, void* data _U_)
     struct ssh_flow_data *global_data=NULL;
     struct ssh_peer_data *peer_data;
 
+    // TODO Retrieve file path after parsing is fixed.
+    // Fixed path for now.
+    const gchar* file_path = "/home/evan/openssh-dump/special_dump.txt"; //getenv("WS_SSH_DECRYPTION_FILE");
+    guchar* key0 = NULL;
+    guchar* key1 = NULL;
+    guint len0 = 0;
+    guint len1 = 0;
+
+    if (file_path != NULL) {
+        parse_ssh_file(file_path, &key0, &len0, &key1, &len1);
+    }
     conversation = find_or_create_conversation(pinfo);
 
     global_data = (struct ssh_flow_data *)conversation_get_proto_data(conversation, proto_ssh);
+    
     if (!global_data) {
         global_data = (struct ssh_flow_data *)wmem_alloc0(wmem_file_scope(), sizeof(struct ssh_flow_data));
+        global_data->current_seqnr = 0;
         global_data->version=SSH_VERSION_UNKNOWN;
         global_data->kex_specific_dissector=ssh_dissect_kex_dh;
+        global_data->key0 = key0;
+        global_data->key1 = key1;
+        global_data->key0_len = len0;
+        global_data->key1_len = len1;
         global_data->peer_data[CLIENT_PEER_DATA].mac_length=-1;
         global_data->peer_data[SERVER_PEER_DATA].mac_length=-1;
 
+        g_print("key0 found: %u with len\n", len0);
+        g_print("key1 found: %u with len\n", len1);
+
         conversation_add_proto_data(conversation, proto_ssh, global_data);
+    } else {
+        // TODO: Where should this go?
+        global_data->current_seqnr = global_data->current_seqnr + 1;
```

Comment:
This should be added to the point where you are sure that the packet is in fact an encrypted packet, and not the initial version negotiation for example. Ensure it is only incremented/used on the first sequential pass (`if (PINFO_FD_VISITED(pinfo)) { /* now you can use and update the sequence number */ }`).

Reply:
Thanks, I think we start incrementing prior to encryption though...
  > The sequence_number is initialized to zero for the first packet, and is
   incremented after every packet (regardless of whether encryption or
   MAC is in use).  It is never reset, even if keys/algorithms are
   renegotiated later.  It wraps around to zero after every 2^32
   packets.  The packet sequence_number itself is not included in the
   packet sent over the wire.

from https://tools.ietf.org/html/rfc4253#section-6.4

and on another note apparently we should be using a uint32 :+1: